autoload/clang_format.vim

```diff
@@ -291,6 +311,92 @@ endfunction
 function! clang_format#disable_auto_format() abort
     let g:clang_format#auto_format = 0
 endfunction
+
+" s:strip: helper function to strip a string
+function! s:strip(string)
+    return substitute(a:string, '^\s*\(.\{-}\)\s*\r\=\n\=$', '\1', '')
+endfunction
+
+" clang_format#get_git_diff
+" a:file must be an absolute path to the file to be processed
+" this function compares the current buffer content against the
+" git index content of the file.
+" this function returns a list of pair of ranges if the file is tracked
+" and has changes, an empty list otherwise
+function! clang_format#get_git_diff(cur_file)
+    let file_path = isdirectory(a:cur_file) ? a:cur_file :
+                \ fnamemodify(a:cur_file, ":h")
+    let top_dir=s:strip(system(
+                \ g:clang_format#git." -C '".file_path."' ".
+                \ "rev-parse --show-toplevel"))
+    if v:shell_error != 0
+        return []
+    endif
+    let cur_file = s:strip(s:system(
+                \ g:clang_format#git." -C '".top_dir."' ".
+                \ "ls-files --error-unmatch '".a:cur_file."'"))
+    if v:shell_error != 0
+        return []
+    endif
+    let source = join(getline(1, '$'), "\n")
+    " git show :file shows the staged content of the file:
+    "  - content in index if any (staged but not commmited)
+    "  - else content in HEAD
+    " this solution also solves the problem for 'git mv'ed file:
+    "  - if the current buffer has been renamed by simple mv (without git
+    "    add), the file is considered as untracked
+    "  - if the renamed file has been git added or git mv, git show :file
+    "    will show the expected content.
+    " this barbarian command does the following:
+    "  - diff --*-group-* options will return ranges (start,end) for each
+    "    diff chunk
+    "  - <(git show :file) is a process substitution, using /dev/fd/<n> as
+    "    temporary file for the output
+    "  - - is stdin, which is current buffer content in variable 'source'
+    let diff_cmd =
+            \ 'diff <('.g:clang_format#git.' show :"'.cur_file.'") - '.
+            \ '--old-group-format="" --unchanged-group-format="" '.
+            \ '--new-group-format="%dF-%dL%c''\\012''" '.
+            \ '--changed-group-format="%dF-%dL%c''\\012''"'
+    let ranges = s:system(diff_cmd, source)
+    if !(v:shell_error == 0 || v:shell_error == 1)
+        echoerr diff_cmd
+        echoerr ranges
+        throw "clang-format: git diff failed"
```

Comment:
```suggestion
        throw printf("clang-format: git diff failed `%s` for ranges %s", diff_cmd, ranges)
```

Please include all information in error message. Consider the case where this function is used by other plugin.